iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnection.java

```diff
@@ -307,6 +362,29 @@ public void close()
 
             log.debug("Closing MQTT connection");
 
+            // When connecting to E4K, the client is expected to publish a message before gracefully closing the connection
+            // so that E4K knows to end the upstream connection for this device. Note that the MQTT connection's "will"
+            // message isn't sufficient to handle this case because the "will" message is only sent on ungraceful disconnections.
+            if (this.config.isConnectingToMqttGateway())
+            {
+                try
+                {
+                    IMqttDeliveryToken token = this.mqttAsyncClient.publish(
+                        E4K_CONNECTION_CLOSE_TOPIC,
+                        E4K_CONNECTION_CLOSE_MESSAGE_PAYLOAD,
+                        E4K_CONNECTION_CLOSE_MESSAGE_QOS,
+                        E4K_CONNECTION_CLOSE_MESSAGE_RETAINED);
+
+                    // If the message fails to be sent/acknowledged in this time span, an MqttException will be thrown here
+                    token.waitForCompletion(E4K_CONNECTION_CLOSE_MESSAGE_TIMEOUT_MILLISECONDS);
+                }
+                catch (MqttException e)
+                {
+                    log.warn("Failed to send the \"Connection will be closed gracefully\" message to the MQTT gateway. " +
```

Comment:
This is a bit of an odd case, but it seems likely that this catch block would only happen if the connection was ungracefully dropped, so the will message should cover for us here.